File: backend/src/modules/payments/providers/nagad.adapter.ts
import { randomUUID } from 'node:crypto';

import { mockInitiation, mockReceiptFromRequest, mockStatus } from './mock.helpers.js';
import type {
  PaymentInitiation,
  PaymentReceipt,
  PaymentRequest,
  PaymentStatus,
  ProviderAdapter,
} from './provider.interface.js';

/**
 * Mock Nagad adapter.
 *
 * Merchant API docs are gated; this adapter returns deterministic mock
 * responses conforming to the same DTOs as the live adapter will. Swapping
 * to the live implementation requires zero changes to callers.
 * See RISK_REGISTER.md R-PRV-01.
 */
export class NagadAdapter implements ProviderAdapter {
  readonly id = 'nagad' as const;

  async initiate(req: PaymentRequest): Promise<PaymentInitiation> {
    if (req.type === 'MERCHANT_PAYMENT') {
      return mockInitiation(req, {
        providerTxnId: `NAGAD-M-${randomUUID()}`,
        redirectUrl: `https://sandbox.mynagad.com:10061/api/dfs-service/mock/checkout?id=${req.idempotencyKey}`,
      });
    }
    return this.launchDialerPassThrough(req);
  }

  launchDialerPassThrough(req: PaymentRequest): Promise<PaymentInitiation> {
    const ussd = `*167*1*${req.recipientPhone ?? ''}*${Math.floor(req.amountMinorUnits / 100)}#`;
    return Promise.resolve(
      mockInitiation(req, {
        providerTxnId: `NAGAD-D-${randomUUID()}`,
        ussdString: ussd,
      }),
    );
  }

  pollStatus(providerTxnId: string): Promise<PaymentStatus> {
    return Promise.resolve(mockStatus(providerTxnId, 'SUCCESS'));
  }

  fetchReceipt(providerTxnId: string): Promise<PaymentReceipt> {
    return Promise.resolve(
      mockReceiptFromRequest(
        {
          providerId: 'nagad',
          type: 'MERCHANT_PAYMENT',
          amountMinorUnits: 0,
          currency: 'BDT',
          idempotencyKey: '',
        },
        providerTxnId,
      ),
    );
  }
}

File: mobile/lib/core/providers/nagad_adapter.dart
import 'payment_models.dart';
import 'provider_adapter.dart';
import 'provider_id.dart';

/// Mock Nagad adapter — returns deterministic responses conforming to the
/// same DTOs the live adapter will. Swap to live endpoints when merchant
/// credentials are onboarded (RISK_REGISTER.md R-PRV-01).
class NagadAdapter implements ProviderAdapter {
  @override
  ProviderId get id => ProviderId.nagad;

  @override
  Future<PaymentInitiation> initiate(PaymentRequest req) async {
    if (req.type == 'MERCHANT_PAYMENT') {
      return PaymentInitiation(
        providerTxnId: 'NAGAD-M-${DateTime.now().millisecondsSinceEpoch}',
        redirectUrl: 'https://sandbox.mynagad.com/mock?id=${req.idempotencyKey}',
        status: 'INITIATED',
      );
    }
    return launchDialerPassThrough(req);
  }

  @override
  Future<PaymentInitiation> launchDialerPassThrough(PaymentRequest req) async {
    final taka = req.amountMinorUnits ~/ 100;
    final ussd = '*167*1*${req.recipientPhone ?? ''}*$taka#';
    return PaymentInitiation(
      providerTxnId: 'NAGAD-D-${DateTime.now().millisecondsSinceEpoch}',
      status: 'INITIATED',
      instructions: PaymentInstructions(method: 'DIALER_PASS_THROUGH', ussdString: ussd),
    );
  }

  @override
  Future<PaymentStatus> pollStatus(String providerTxnId) async => PaymentStatus(
        providerTxnId: providerTxnId,
        status: 'SUCCESS',
        updatedAt: DateTime.now().toIso8601String(),
      );

  @override
  Future<PaymentReceipt> fetchReceipt(String providerTxnId) async => PaymentReceipt(
        providerTxnId: providerTxnId,
        providerId: ProviderId.nagad,
        type: 'MERCHANT_PAYMENT',
        status: 'SUCCESS',
        amountMinorUnits: 0,
        currency: 'BDT',
      );
}
